### ISSUE TITLE
Integrate jupyterhub-idle-culler with KubeIngressProxy

### ISSUE THREAD

**Issue by droctothorpe** [jupyterhub/jupyterhub-idle-culler#42]
### Proposed change
> last_activity data for routes will be available when using configurable-http-proxy as JupyterHub does by default, but if for example traefik-proxy is used as it is in the TLJH distribution, no such data will be available.

We use JupyterHub + [KubeIngressProxy](https://github.com/jupyterhub/kubespawner/blob/main/kubespawner/proxy.py#L46) + the nginx ingress controller. Based on the docs, I suspect that the `last_activity` is never updated, so notebooks are culled even when they're active.

Are there any plans to integrate these two capabilities? If not, would you be open to a PR to that effect (assuming it's possible)?

Thanks!

**Comment by manics** [comment]
The idle culler is agnostic to the underlying spawner. Assuming you're right about the cause it sounds like the fix needs to be in KubeIngressProxy not this repo. Is that your understanding? If so I'll transfer the issue to KubeSpawner.

**Comment by droctothorpe** [comment]
Thanks for getting back to me, @manics! 

KIP just creates the ingress resources. The nginx ingress controller is what's actually responsible for managing the proxy, and it's completely oblivious to the concept of `last_activity`, since that's a very JH / configurable-http-proxy convention. 

I guess what I'm wondering if it makes sense to either modify jupyterhub-idle-culler or create a discrete culler that's more generic / isn't contingent on the use of configurable-http-proxy.

Some approaches that come to mind:
1. Evaluating idleness using a daemon on the notebook container that terminates the container locally instead of relying on a centralized service.
2. Determining idleness by checking HTTP requests frequency via prometheus. 

Totally understand if this is out of scope. We're pretty unique in our adoption of KIP I suspect.

**Comment by droctothorpe** [comment]
Hm, alternatively, why not hit the /api/status endpoints of the notebooks? I think that's what KF does: https://github.com/kubeflow/kubeflow/blob/master/components/notebook-controller/pkg/culler/culler.go.

**Comment by manics** [comment]
There are some past discussions on supporting plugins or configurable hooks:
- https://github.com/jupyterhub/jupyterhub-idle-culler/issues/25
- https://github.com/jupyterhub/jupyterhub-idle-culler/issues/9
- https://github.com/jupyterhub/jupyterhub-idle-culler/issues/8

No progress yet, but maybe your use-case could be a driver?

If you haven't already seen https://discourse.jupyter.org/t/which-is-the-correct-way-to-cull-idle-kernels-and-notebook/8123/7 it explains some of the complexity of idle detection?

Did you see https://discourse.jupyter.org/t/jupyterhub-amalthea/12208 (posted today)? Might be worth seeing how they implemented idle culling.
